src/formats/inp.rs

```diff
@@ -31,37 +15,69 @@ const TEX: &[u8] = b"TEX_SECT";
 /// Extended section header
 // const EXT: &[u8] = b"EXT_SECT";
 
-fn be_u32_plus_1(i: &[u8]) -> IResult<&[u8], u32> {
-    let (i, int) = be_u32(i)?;
-    Ok((i, 1 + int))
+fn read_u8<R: io::Read>(reader: &mut R) -> io::Result<u8> {
+    let mut buf = [0_u8; 1];
+    reader.read_exact(&mut buf)?;
+    Ok(buf[0])
 }
 
-/// Parse a `.inp` Inochi Puppet from memory.
-pub fn parse_inp(i: &[u8]) -> IResult<&[u8], Model> {
-    let (i, _) = tag(MAGIC)(i)?;
-    let (i, json_payload) = length_data(be_u32)(i)?;
+fn read_be_u32<R: io::Read>(reader: &mut R) -> io::Result<u32> {
+    let mut buf = [0_u8; 4];
+    reader.read_exact(&mut buf)?;
+    Ok(u32::from_be_bytes(buf))
+}
 
-    let (i, _) = tag(TEX)(i)?;
-    let (mut i, num_textures) = be_u32(i)?;
-    let mut textures = Vec::new();
-    for _ in 0..num_textures {
-        let (i2, texture) = length_value(be_u32_plus_1, parse_texture)(i)?;
-        textures.push(texture);
-        i = i2;
+fn read_array<R: io::Read, const N: usize>(reader: &mut R) -> io::Result<[u8; N]> {
+    let mut data = [0_u8; N];
+    reader.read_exact(&mut data)?;
+    Ok(data)
+}
+
+fn read_vec<R: io::Read>(reader: &mut R, length: u32) -> io::Result<Vec<u8>> {
+    let length = length as usize;
+    let mut data: Vec<MaybeUninit<u8>> = Vec::with_capacity(length);
+    unsafe { data.set_len(length) };
+    let mut data: Vec<u8> = unsafe { std::mem::transmute(data) };
+    reader.read_exact(&mut data)?;
+    Ok(data)
```

Comment:
Casual unsafety isn't something you want here.

There is no guaranties that ``Vec<MaybeUninit<u8>>`` have the same layout as ``Vec<u8>``.

```suggestion
    let length = length as usize;
    let mut data: Vec<u8> = Vec::with_capacity(length);
    reader.read_exact(&mut data)?;
    Ok(data)
```

Reply:
> There is no guaranties that `Vec<MaybeUninit<u8>>` have the same layout as `Vec<u8>`.

It is guaranteed that `MaybeUninit<u8>` has the same layout as `u8`: https://doc.rust-lang.org/std/mem/union.MaybeUninit.html#layout-1
Although you may be right about the `Vec` containing it.

Your solution won't work because `read_exact` reads using the length of the vector and not its capacity.
Someone benchmarked it and it seems that initializing the vector with zeros costs no performance loss over this unsafe solution, so I'm gonna do that instead.